power_grid_model_c/power_grid_model/include/power_grid_model/math_solver/sparse_lu_solver.hpp

```diff
@@ -328,6 +405,210 @@ template <class Tensor, class RHSVector, class XVector> class SparseLUSolver {
+        rhs_ = rhs;
+        // initialize x to zero
+        for (Idx row = 0; row != size_; ++row) {
+            x[row] = XVector{};
+        }
+        // initialize residual to rhs
+        // because r = b - A * x = b - 0 = b
+        residual_ = rhs_;
+        // initialize dx to zero
+        // pre-allocate memory
+        dx_ = x;
+    }
+
+    void calculate_residual(std::vector<XVector> const& x) {
+        auto const& row_indptr = *row_indptr_;
+        auto const& col_indices = *col_indices_;
+        auto const& original_matrix = original_matrix_.value();
+        auto const& rhs = rhs_.value();
+        auto& residual = residual_.value();
+        // calculate residual
+        for (Idx row = 0; row != size_; ++row) {
+            residual[row] = rhs[row];
+            // loop all columns
+            for (Idx idx = row_indptr[row]; idx != row_indptr[row + 1]; ++idx) {
+                // subtract
+                residual[row] -= dot(original_matrix[idx], x[col_indices[idx]]);
+            }
+        }
+    }
+
+    double iterate_and_backward_error(std::vector<XVector>& x) {
+        double backward_error_max_demoniator = 0.0;
+        double backward_error_max_numerator = 0.0;
+        auto const& row_indptr = *row_indptr_;
+        auto const& col_indices = *col_indices_;
+        auto const& original_matrix = original_matrix_.value();
+        auto const& rhs = rhs_.value();
+        auto const& residual = residual_.value();
+        auto const& dx = dx_.value();
+        using RealValueType = std::conditional_t<is_block, Eigen::Array<double, block_size, 1>, double>;
+        // calculate backward error and then iterate x
+        for (Idx row = 0; row != size_; ++row) {
+            // error denominator by |rhs|
+            RealValueType denominator = cabs(rhs[row]);
+            // then append |A| * |x|
+            for (Idx idx = row_indptr[row]; idx != row_indptr[row + 1]; ++idx) {
+                denominator += dot(cabs(original_matrix[idx]), cabs(x[col_indices[idx]]));
+            }
+            RealValueType const numerator = cabs(residual[row]);
+            backward_error_max_demoniator = std::max(backward_error_max_demoniator, max_val(denominator));
+            backward_error_max_numerator = std::max(backward_error_max_numerator, max_val(numerator));
+            // iterate x
+            x[row] += dx[row];
+        }
+        // then |r|_max / (|rhs| + |A| * |x|)_max
+        return backward_error_max_numerator / backward_error_max_demoniator;
+    }
+
+    void initialize_pivot_perturbation(std::vector<Tensor> const& data) {
+        // save a copy of original matrix
+        original_matrix_ = data;
+        // calculate the block-wise non-diagonal infinite norm of the matrix
+        // that is:
+        // 1. calculate the infinite norm of each individual block
+        // 2. sum all norms of the blocks per row, except the diagonal block
+        // 3. take the maximum of all the sums
```

Comment:
i guess it's approximately good enough, but this is not an exact infinity norm:

take the following example:

$$
\begin{pmatrix}
\begin{pmatrix}
1 && 0 \\
0 && 3
\end{pmatrix} && \begin{pmatrix}
3 && 0 \\
0 && 0
\end{pmatrix}
\end{pmatrix}
$$

the infinity norm should be the abs sum of the 1st row: `1 + 3 = 4`, but the infinity norm of the block-wise infinity norms says that it's `3 + 3 = 6`, which is larger.

should we stick to the existing solution or change it to the `rowwise().sum().maxCoeff()` of the sums of the absolute values of the blocks?

Reply:
That what I call it block-wise off-diagonal infinite norm.